01-getting-started/projects/final/app/src/main/java/com/raywenderlich/podplay/service/PodplayMediaCallback.kt

```diff
@@ -0,0 +1,308 @@
+      }
+    } else {
+      audioManager.abandonAudioFocus(null)
+    }
+  }
+
+  private fun initializeMediaPlayer() {
+    // TODO 4
+
+    if (mediaPlayer == null) {
+      mediaPlayer = MediaPlayer()
+      mediaPlayer!!.setOnCompletionListener {
+        setState(PlaybackStateCompat.STATE_PAUSED)
+      }
+      mediaNeedsPrepare = true
+    }
+  }
+
+  private fun changeSpeed(extras: Bundle) {
+    var playbackState = PlaybackStateCompat.STATE_PAUSED
+    if (mediaSession.controller.playbackState != null) {
+      playbackState = mediaSession.controller.playbackState.state
+    }
+    setState(playbackState, extras.getFloat(CMD_EXTRA_SPEED))
+  }
+
+  private fun setState(state: Int, newSpeed: Float? = null) {
+    var position: Long = -1
+
+    mediaPlayer?.let {
+      position = it.currentPosition.toLong()
+    }
+
+    var speed = 1.0f
+
+    speed = newSpeed ?: (mediaPlayer?.playbackParams?.speed ?: 1.0f)
+    mediaPlayer?.let { mediaPlayer ->
+      try {
+        mediaPlayer.playbackParams = mediaPlayer.playbackParams.setSpeed(speed)
+      } catch (e: Exception) {
+
+        mediaPlayer.reset()
+        mediaUri?.let { mediaUri ->
+          mediaPlayer.setDataSource(context, mediaUri)
+        }
+        mediaPlayer.prepare()
+
+        mediaPlayer.playbackParams = mediaPlayer.playbackParams.setSpeed(speed)
+        mediaPlayer.seekTo(position.toInt())
+
+        if (state == PlaybackStateCompat.STATE_PLAYING) {
+          mediaPlayer.start()
+        }
+      }
+    }
+
+    val playbackState = PlaybackStateCompat.Builder()
+        .setActions(
+            PlaybackStateCompat.ACTION_PLAY or
+                PlaybackStateCompat.ACTION_STOP or
+                PlaybackStateCompat.ACTION_PLAY_PAUSE or
+                PlaybackStateCompat.ACTION_PAUSE)
+        .setState(state, position, speed)
+        .build()
+
+    mediaSession.setPlaybackState(playbackState)
+
+    if (state == PlaybackStateCompat.STATE_PAUSED ||
+        state == PlaybackStateCompat.STATE_PLAYING) {
+      listener?.onStateChanged()
+    }
+  }
+
+  private fun prepareMedia() {
+    // TODO 5
+
+    if (newMedia) {
+      newMedia = false
+      mediaPlayer?.let { mediaPlayer ->
+        mediaUri?.let { mediaUri ->
+          if (mediaNeedsPrepare) {
+            mediaPlayer.reset()
+            mediaPlayer.setDataSource(context, mediaUri)
+            mediaPlayer.prepare()
+          }
+          mediaExtras?.let { mediaExtras ->
+            mediaSession.setMetadata(MediaMetadataCompat.Builder()
+                .putString(MediaMetadataCompat.METADATA_KEY_TITLE,
+                    mediaExtras.getString(MediaMetadataCompat.METADATA_KEY_TITLE))
+                .putString(MediaMetadataCompat.METADATA_KEY_ARTIST,
+                    mediaExtras.getString(MediaMetadataCompat.METADATA_KEY_ARTIST))
+                .putString(MediaMetadataCompat.METADATA_KEY_ALBUM_ART_URI,
+                    mediaExtras.getString(
```

Comment:
formatting

Reply:
Will update with RW code formatting style.

Reply:
https://trello.com/c/gKEwBixK/25-apply-rw-style-format-throughout-the-project